### ISSUE TITLE
input a video instead of webcam

### ISSUE THREAD

**Issue by amir-za** [homuler/MediaPipeUnityPlugin#176]
hello. 
I was wondering if it's possible to extract landmarks from a video instead of a webcam. so I added a video player component to WebCamScreen. 
and wrote the below code in WebCamScreenController in order to read each frame of the video.

```csharp
IEnumerator playVideo()
{
    videoPlayer.playOnAwake = false;
    videoPlayer.Stop();
    videoPlayer.renderMode = VideoRenderMode.APIOnly;
    videoPlayer.Prepare();
    while (!videoPlayer.isPrepared)
    {
        Debug.Log("Preparing Video");
        yield return null;
    }
    videoPlayer.sendFrameReadyEvents = true;
    videoPlayer.frameReady += get_video_texture;
    videoPlayer.Play();

    TextureFramePool.Instance.SetDimension((int)videoPlayer.width, (int)videoPlayer.height);
    vid_scr = new Texture2D((int)videoPlayer.width, (int)videoPlayer.height, TextureFormat.RGBA32, false); // a texture2d to store video frames
    Debug.Log("Playing Video");
    while (videoPlayer.isPlaying)
    {
        Debug.LogWarning("Video Time: " + Mathf.FloorToInt((float)videoPlayer.time));
        yield return new WaitForSeconds(0.5f);

    }
    Debug.Log("Done Playing Video");
}
```

```csharp
public void get_video_texture(VideoPlayer vp, long frameIndex)
{
    // check to see if go to next frame according to framerate
    bool if_update = (float)frameIndex / vp.frameRate >= Time.realtimeSinceStartup;
    if ((long)vp.frame < (long)vp.frameCount && if_update)
        vp.frame = frameIndex + 1;
}
```

and i added StartCoroutine(playVideo());   in start method.

also i changed RequestNextFrame in WebCamScreenController.cs to this :
```csharp
public TextureFramePool.TextureFrameRequest RequestNextFrame()
{
    return TextureFramePool.Instance.RequestNextTextureFrame((TextureFrame textureFrame) => {
            if (videoPlayer.isPlaying && videoPlayer.frame > 0)
            {
                Graphics.CopyTexture(videoPlayer.texture, vid_scr);
                textureFrame.CopyTextureFrom(vid_scr);
            }                    
    });
}
```
and changed CopyTextureFrom method to this
```csharp
public void CopyTextureFrom(Texture2D src)
 {
     Graphics.CopyTexture(src, texture);
 }
```
now it doesn't give me any errors but when I call RenderOutput to get NormalizedLandmarkList it returns empty.
I think the problem is in PushInput when it tries to create an ImageFrame from nextFrame. in the below line. it returns an empty ImageFrame.
```csharp
var imageFrame = new ImageFrame(
          ImageFormat.Format.SRGBA, textureFrame.width, textureFrame.height, 4 * textureFrame.width, textureFrame.GetRawNativeByteArray());
```

I don't know if there is an easier way to do this. 
sorry if it got lengthy and also poor English.

**Comment by homuler** [comment]
> it returns an empty ImageFrame.

What do you mean by 'empty'? Does `imageFrame.width` and `imageFrame.height` equal zero or `imageFrame.IsEmpty()` return true?

**Comment by amir-za** [comment]
> 
> What do you mean by 'empty'? Do `imageFrame.Width()` and `imageFrame.Height()` equal zero or does `imageFrame.IsEmpty()` return true?

imageFrame.Width() and imageFrame.Height() returns correct values and imageFrame.IsEmpty() also returns false but when i call ` DrawScreen(ImageFrame imageFrame)` to draw the imageFrame on the WebCamScreen, it doesn't show anything. 
but when input is from the webcam, calling ` DrawScreen(ImageFrame imageFrame)` works fine.

**Comment by homuler** [comment]
> ```cs
> public TextureFramePool.TextureFrameRequest RequestNextFrame()
> {
>     return TextureFramePool.Instance.RequestNextTextureFrame((TextureFrame textureFrame) => {
>             if (videoPlayer.isPlaying && videoPlayer.frame > 0)
>             {
>                 Graphics.CopyTexture(videoPlayer.texture, vid_scr);
>                 textureFrame.CopyTextureFrom(vid_scr);
>             }                    
>     });
> }
> ```

Does `videoPalyer.texture` contain expected images and the condition `videoPlayer.isPlaying && vidoPlayer.frame > 0` surely become true?
If `textureFrame.GetRawNativeByteArray()` returns a native pointer to the byte array that contains the video frame, I think you can use `VideoPlayer` instead of `WebCamTexture`.

**Comment by amir-za** [comment]
> Does videoPalyer.texture contain expected images and the condition videoPlayer.isPlaying && vidoPlayer.frame > 0 surely become true?

The condition becomes true for sure.
but I checked the `textureFrame.GetRawNativeByteArray() ` and you were right. it didn't contain true bytes from the video frame. actually, `videoPalyer.texture` is of type 'Texture' and `vid_scr ` is a `Textue2D`, and when I use CopyTexture to store the `Texture` to `Texture2D` the pixels data of `Texture2D` didn't change. 
the only solution that I came up with was using a `UI.RawImage` that played the video in the corner of the screen and using `Texture2D.ReadPixels()` to read the pixel data from the video. this works but the video needs to play on the screen.
